fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanSchedularServiceImpl.java

```diff
@@ -184,4 +191,133 @@ public void recalculateInterest() throws JobExecutionException {
 
 	}
 
-}
+	@Override
+	@CronTarget(jobName = JobName.RECALCULATE_INTEREST_FOR_LOAN)
+	public void recalculateInterest(Map<String, String> jobParameters) {
+		//gets the officeId
+		final String officeId = jobParameters.get("officeId");
+		logger.info(officeId);
+		Long officeIdLong=Long.valueOf(officeId);
+		//gets the Office object
+		final Office office = this.officeRepository.findOne(officeIdLong);
+		if(office == null)
+			throw new OfficeNotFoundException(officeIdLong);
+		final int threadPoolSize=Integer.parseInt(jobParameters.get("thread-pool-size"));
+		final int batchSize=Integer.parseInt(jobParameters.get("batch-size"));
+
+		recalculateInterest(office,threadPoolSize,batchSize);
+	}
+
+	@Override
+	public void recalculateInterest(Office office, int threadPoolSize, int batchSize) {
+		final int pageSize = batchSize * threadPoolSize;
+
+		//initialise the executor service with fetched configurations
+		final ExecutorService executorService = Executors.newFixedThreadPool(threadPoolSize);
+
+		Long maxLoanIdInList = 0L;
+		final String officeHierarchy = office.getHierarchy() + "%";
+
+		//Get the loanIds from service
+		List<Long> loanIds = this.loanReadPlatformService
+				.fetchLoansForInterestRecalculation(pageSize, maxLoanIdInList, officeHierarchy);
+
+		// gets the loanIds data set iteratively and call addAccuruals for that paginated dataset
+		do {
+			int totalFilteredRecords = loanIds.size();
+			logger.info("Starting accrual - total filtered records - " + totalFilteredRecords);
+			recalculateInterest(loanIds, threadPoolSize, batchSize,
+					executorService);
+			maxLoanIdInList+=pageSize+1;
+			loanIds = this.loanReadPlatformService
+					.fetchLoansForInterestRecalculation(pageSize, maxLoanIdInList, officeHierarchy);
+		} while (!CollectionUtils.isEmpty(loanIds));
```

Comment:
are you sure that this is properly thread safe? Hopefully the `List<Long> loanIds` is not just a normal `ArrayList` or so, because then it's not.. it would have to be a something like a `CopyOnWriteArrayList` or `ConcurrentSkipListSet` or so for this to be thread safe. Ihave mot looked more into it, just wanted to point it out for you.

Reply:
make sure that `List<Long> loanIds` is something like `Collections.synchronizedList(new ArrayList())` should be fine as well.

Reply:
Any comment re. this? Do you think this code is thread safe as it is?

Reply:
@vorburger I just made them thread-safe. Thanks for your suggestions